### PR TITLE
FIX: Add order to outputted stylesheet link tags

### PR DESCRIPTION
**Pull request by pmusaraj**
Theme and component stylesheets have so far been loaded without a specified order. This causes issues. Especially on sites with multiple themes and multiple components, styles from components would sometimes take precedence over the base theme and sometimes it would be the other way around. 

This PR orders the theme stylesheets thusly: 

- remote theme components (ordered alphabetically)
- remote main theme (if applicable)
- local theme components (ordered alphabetically)
- local main theme (if applicable)

Note that stylesheet ordering is **not yet** enabled by default, for now this behaviour is behind a hidden site setting: `order_stylesheets`. 

To use this, do the following in the Rails console: 

```bash
pry(main)> SiteSetting.order_stylesheets = true

pry(main)> Stylesheet::Manager.clear_theme_cache!
```

This PR also removes an extra layer of caching. We cache the full array of stylesheets for a specific target, and we also cached each stylesheet path, which is redundant.

### CODE REVIEW

**Comment by eviltrout** [comment]
This looks good to me but could it possibly break existing themes where they depended on the previous order? I agree this order is better, for the record. I just want to figure out if any forums will be affected negatively.

**Comment by pmusaraj** [comment]
Yes, it can potentially break existing sites. We currently don't have a load order, though, so this would have caused issues sooner or later anyway... 

I will test this on a few sites with many themes and components before merging. Also, the main risk here are sites that have styles in a component that are supposed to override styles in the parent theme.

**Comment by eviltrout** [comment]
Thanks I think it's better to have an order than not, but we should be aware of the risk and have a plan of action when merging.

File: spec/components/stylesheet/manager_spec.rb [hunk 1]
```diff
@@ -135,6 +135,67 @@ def manager(theme_id = nil)
       )
     end
 
+    context "stylesheet order " do
+      let(:z_child_theme) { Fabricate(:theme, component: true, name: "ze component").tap { |z|
+        z.set_field(target: :desktop, name: "scss", value: ".child_desktop{.scss{color: red;}}")
+        z.save!
+      }}
```
Comment (tgxworld) [inline spec/components/stylesheet/manager_spec.rb:142]:
```suggestion
      let(:z_child_theme) do 
        Fabricate(:theme, component: true, name: "ze component").tap do |z|
          z.set_field(target: :desktop, name: "scss", value: ".child_desktop{.scss{color: red;}}")
          z.save!
        end
      end
```

We usually prefer `do..end` for multi-line blocks

File: spec/components/stylesheet/manager_spec.rb [hunk 2]
```diff
@@ -135,6 +135,67 @@ def manager(theme_id = nil)
...
+
+      let(:remote) { RemoteTheme.create!(remote_url: "https://github.com/org/remote-theme1") }
+
+      let(:child_remote) { Fabricate(:theme, remote_theme: remote, component: true).tap { |t|
+        t.set_field(target: :desktop, name: "scss", value: ".child_desktop{.scss{color: red;}}")
+        t.save!
+      }}
```
Comment (tgxworld) [inline spec/components/stylesheet/manager_spec.rb:149]:
```suggestion
      let(:child_remote) do 
        Fabricate(:theme, remote_theme: remote, component: true).tap do |t|
          t.set_field(target: :desktop, name: "scss", value: ".child_desktop{.scss{color: red;}}")
          t.save!
        end
      end
```

File: lib/stylesheet/manager.rb [hunk 3]
```diff
@@ -211,54 +212,33 @@ def stylesheet_details(target = :desktop, media = 'all')
...
+          href = builder.stylesheet_path(current_hostname)
 
           data[:new_href] = href
           stylesheets << data
         end
+
+        if stylesheets.size > 1
+          stylesheets = stylesheets.sort_by { |s|
+            [s[:remote] ? 0 : 1, s[:theme_id] == @theme_id ? 1 : 0, s[:theme_name]]
+          }
```
Comment (tgxworld) [inline lib/stylesheet/manager.rb:235]:
```suggestion
          stylesheets = stylesheets.sort_by do |s|
            [
              s[:remote] ? 0 : 1, 
              s[:theme_id] == @theme_id ? 1 : 0, 
              s[:theme_name]
            ]
          end
```

Minor but I find this easier to read as compared to all the code being in a single line.

File: spec/components/stylesheet/manager_spec.rb [hunk 4]
```diff
@@ -135,6 +135,67 @@ def manager(theme_id = nil)
...
+        theme.add_relative_theme!(:child, z_child_theme)
+        theme.add_relative_theme!(:child, child_remote)
+
+        manager = manager(theme.id)
+        hrefs = manager.stylesheet_details(:desktop_theme, 'all')
+
+        parent = hrefs.select { |href| href[:theme_id] == theme.id }.first
+        childA = hrefs.select { |href| href[:theme_id] == child_theme.id }.first
+        childZ = hrefs.select { |href| href[:theme_id] == z_child_theme.id }.first
+        childR = hrefs.select { |href| href[:theme_id] == child_remote.id }.first
```
Comment (tgxworld) [inline spec/components/stylesheet/manager_spec.rb:161]: We're missing camelCase and snake_case for the variables here. Usually we default to snake_case for variable names in Ruby.

File: spec/components/stylesheet/manager_spec.rb [hunk 5]
```diff
@@ -135,6 +135,67 @@ def manager(theme_id = nil)
...
+
+        expect(link_hrefs).to eq([child_remote_R, child_local_A, child_local_Z, parent_local].join("\n").html_safe)
+      end
+
+      it "output remote child, remote parent, local child" do
+        remote2 = RemoteTheme.create!(remote_url: "https://github.com/org/remote-theme2")
+        remote_main_theme = Fabricate(:theme, remote_theme: remote2, name: "remote main").tap { |t|
+          t.set_field(target: :desktop, name: "scss", value: ".el{color: red;}")
+          t.save!
+        }
```
Comment (tgxworld) [inline spec/components/stylesheet/manager_spec.rb:178]:
```suggestion
        remote_main_theme = Fabricate(:theme, remote_theme: remote2, name: "remote main").tap do |t|
          t.set_field(target: :desktop, name: "scss", value: ".el{color: red;}")
          t.save!
        end
```

**Review by tgxworld — APPROVED** [review]
Some minor code style comments but the PR looks good to me otherwise. Regarding how we would deploy this, I recommend just putting the sorting behind a hidden site setting first and then roll it out slowly across our hosting.

**Pull request merged**